File: julia/5.jl
# https://projecteuler.net/problem=5
#
# 2520 is the smallest number that can be divided by each of
# the numbers from 1 to 10 without any remainder.
#
# What is the smallest positive number that is evenly
# divisible by all of the numbers from 1 to 20?

function _gcd(a, b)
    if a < b
        _gcd(b, a)
    else
        remainder = a % b
        remainder == 0 ? b : _gcd(b, remainder)
    end
end

_lcm(a, b) = a ÷ _gcd(a, b) * b

println(reduce(_lcm, 1, 1:20))

File: dart/5.dart
// https://projecteuler.net/problem=5
//
// 2520 is the smallest number that can be divided by each of
// the numbers from 1 to 10 without any remainder.
//
// What is the smallest positive number that is evenly
// divisible by all of the numbers from 1 to 20?


// Calculate greatest common divisor using Euclid's algorithm.
// Both arguments must be integers greater than 0.
int gcd(int a, int b) {
  if (a < b) {
    return gcd(b, a);
  } else if (b == 0) {
    return a;
  } else {
    return gcd (b, a % b);
  }
}

// Calculate least common multiple.
// Both arguments must be integers greater than 0.
int lcm(int a, int b) {
  return (a * b) ~/ gcd(a, b);
}

void main() {
  final result = List<int>.generate(20, (i) => i + 1)
    .reduce(lcm);

  print(result);
}